enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,580 @@
+
+### Writing to the Probe Resource Directly
+
+An Operator can manage its status by writing directly to the `Probe` resource associated with its operator. The [operator-sdk](https://github.com/operator-framework/operator-sdk) project will provide a library that includes a set of functions geared towards managing the status [subresource](https://book-v1.book.kubebuilder.io/basics/status_subresource.html) of the `Probe` CRD introduced by OLM.
+
+The functions provided within the library will detect if the `Probe` CRD is present on the cluster using the Discovery API. If the `Probe` resource exists, the library will provide a set of functions allowing a user to `GET` or `UPDATE` the operator's `Probe`.
+
+The library will also communicate to OLM that the operator is managing the `Probe` resource by setting the `spec.manager` field to `operator`. When the `spec.manager` field is set to `operator`, the library will throw an error if the `spec.ProbeResources` field is not `nil`.
+
+At the time of this enhancement, the operator will only need to manage its `Upgradeable` condition.
+
+#### Example: Operator Managed Probe Resource
+
+Let's consider what the `Probe` resource may look like for the Foo operator:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: Probe
+metadata:
+  name: foo-operator.v1.0.0
+  namespace: operators
+spec:
+  manager: operator  # When set to `operator`, the operator must manage the state of the probe resource.
+status:
+  conditions:
+  - type: Upgradeable # Overall Upgradeable condition
+    status: false     # The operator is not upgradeable
+    reason: "FooProcess"
+    message: "The operator is not upgradeable because FooProcess is running"
+```
+
+In this case, the operator is communicating that OLM should not upgrade the operator at this time. Likewise, the operator could convey that OLM may upgrade the operator with the following `Probe`:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: Probe
+metadata:
+  name: foo-operator.v1.0.0
+  namespace: operators
+spec:
+  manager: operator
+status:
+  conditions:
+  - type: Upgradeable # Overall Upgradeable condition
+    status: True      # The operator is upgradeable
+    reason: "ReadyForUpgrade"
+    message: "The operator is ready to be upgraded"
+```
+
+### Test Plan
+
+OLM's e2e testing suite will be upgraded to handle the usecase:
+
+- If an operator does not opt-in to "OLM Supported Conditions", OLM reverts to existing behavior.
+- If an operator opts-in to the `Upgradeable` "OLM Supported Conditions":
+  - Upgrades are not blocked if no CRs exist.
+  - Upgrades are not blocked if a CR exists but the condition is missing
+  - Upgrades are not blocked if a CR exists, the mapped condition type is present and is not equal to `false`.
+  - Upgrades are blocked if a CR exists, the mapped condition type is present and `true`.
```

Comment:
Something in here is not right, because if the condition is `true`, then both "present and `true`" (blocked) and "present and is not equal to `false`" (not blocked) would match.

Reply:
Great point - this should have been updated to incorporate the boolean expressions over condition types.

Reply:
This is an error - I will correct it. Thanks @wking!

Reply:
Addressed in the current version of the PR.